### PR TITLE
Route autograd.Function under vmap through functorch

### LINKED ISSUE #134784: Dynamo x autograd.Function x funtorch transform: intermediate sliced tensor doesn't require grad caused incorrect result
### 🐛 Describe the bug

Let's use this small repro to demonstrate this issue:
```
import torch
import inspect

@torch.compile(backend="eager", fullgraph=True)
def fn(x):
    x = torch.vmap(gn)(x)
    return x + 1

def gn(x: torch.Tensor) -> torch.Tensor:
    if x.requires_grad:
        return x + 1
    else:
        return x - 1

x = torch.ones(2, 3, requires_grad=True)
print(fn(x))

```
The output:
```
tensor([[1., 1., 1.],
        [1., 1., 1.]], grad_fn=<AddBackward0>)
```
This means the ```gn``` function goes with the ```else``` branch, because of the intermediate sliced tensor doesn't require grad. However, if ```gn``` is an autograd function, then we would check if the input tensor requires grad and decide how to proceed.
https://github.com/pytorch/pytorch/blob/b977abd5de0ea8c37f6100e7fc170830a07ff47b/torch/_dynamo/variables/misc.py#L652-L656

We need to revisit the behavior of how to compile autograd function regarding to ```requires_grad=True/False```.

### Versions

N/A

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186303

Dynamo decided whether to emit the autograd.Function higher-order op from
VariableTracker requires_grad metadata. Inside vmap, BatchedTensor intentionally
reports requires_grad=False, so Dynamo treated autograd.Function.apply as an
inference call and inlined forward directly. That skipped eager's functorch
custom_function_call path, so custom vmap rules and their backward behavior were
not honored under torch.compile.

Fix this by routing autograd.Function.apply through custom_function_call when a
functorch transform is active before falling back to direct forward inlining.
AutogradFunctionVariable now exposes the underlying Function subclass as a
Python constant with attribute access for functorch's custom_function_call
implementation, while preserving the special apply handling.

For setup_context-style Functions, mirror eager Function.apply by binding
forward defaults before the Dynamo kwargs normalization. Defaults inserted by
binding use DefaultsSource, and positional defaults also guard the __defaults__
length so changes to the default-to-parameter mapping invalidate compiled code.
Varargs are preserved when defaults are rebound.

I considered changing BatchedTensor.requires_grad, but that behavior is
intentional and eager vmap already exposes it. The narrower fix is to make the
compiled autograd.Function path use the same functorch dispatch decision as
eager instead of relying on BatchedTensor requires_grad metadata.

Fixes #134784
Generated by my agent

Test Plan:
- python test/dynamo/test_autograd_function.py AutogradFunctionFunctorchTests.test_setup_context_preserves_varargs AutogradFunctionFunctorchTests.test_vmap_custom_rule_binds_tensor_default AutogradFunctionFunctorchTests.test_setup_context_binds_default_args_without_functorch AutogradFunctionFunctorchTests.test_vmap_custom_rule_binds_default_args
- python test/dynamo/test_autograd_function.py -k AutogradFunctionFunctorchTests
- python test/dynamo/test_autograd_function.py
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR and addressed the Dr. CI failures by narrowing the new autograd.Function functorch dispatch to explicit custom vmap rules, adding coverage for a grad-requiring unbatched vmap argument, and adding Dynamo DispatchKeySet binary-op support for nested vmap tracing. Local targeted Dynamo failures and lintrunner now pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the generated-vmap review gap by making the custom autograd.Function path vmap-specific, preserving torch.func.grad-only traces, matching eager errors for old-style/no-vmap Functions, and graph-breaking generated-vmap rules instead of silently inlining forward. I also investigated the current Dr. CI failures: DebertaV2ForMaskedLM fails during HF import/manual_seed setup, and the TorchTitan failure matches the known torchcomms process-group issue in pytorch/pytorch#186031, so neither appears caused by this PR.

**Comment by jansel** [comment]
Duplicate of #186330, which fixes the same root cause — compiled `autograd.Function` subclasses with a manual vmap rule (`generate_vmap_rule=False`) routed through functorch's `custom_function_call` — using the same rerouting helpers. #186330 keeps the change focused on that fix; this PR additionally bundles an unrelated DispatchKeySet binop change that is failing CI broadly. The consolidated #186330 now carries `Fixes #133493` and `Fixes #134784`.
